src/components/Dropdown/DropdownList.tsx

```diff
@@ -56,21 +57,98 @@ export const DropdownList = (props: IDropdownListProps): ReactElement => {
     ],
   });
 
+  const close = () => {
+    if (visible) {
+      setVisible(false);
+      onClosed();
+    }
+  };
+
+  const open = () => {
+    if (!visible) {
+      setVisible(true);
+      onExpanded();
+    }
+  };
+
+  /* Click on dropdown */
   const handleClick = () => {
     setVisible(!visible);
+    visible ? onExpanded() : onClosed();
   };
 
-  const handleSelect = useCallback(
-    (item: ItemType) => {
-      return () => {
-        onSelect(item.id);
-      };
-    },
-    [onSelect],
-  );
+  /* keydown on dropdown */
+  const handleKeyDown = (e: KeyboardEvent) => {
+    switch (e.key) {
+      case 'Enter':
+        e.preventDefault();
+        handleClick();
+        return;
+      case 'Space':
+        e.preventDefault();
+        handleClick();
+        return;
+      case 'Escape':
+        close();
+        return;
+      case 'ArrowDown':
+        e.preventDefault();
+        open();
+        focusItem(0);
+        return;
+      default:
+        return;
+    }
+  };
+
+  /* selected item */
+  const handleSelect = (item: ItemType) => {
+    onSelect(item.id);
+    close();    
+  };
+
+  /* keydown on item */
+  const handleItemKeyDown = (e: KeyboardEvent, item: ItemType, index: number) => {
+    switch (e.key) {
+      case 'Enter':
+        e.preventDefault();
+        handleSelect(item);
+        return;
+      case 'Space':
+        e.preventDefault();
+        handleSelect(item);
+        return;
+      case 'ArrowUp':
+        e.preventDefault();
+        focusItem(index - 1);
+        return;
+      case 'ArrowDown':
+        e.preventDefault();
+        focusItem(index + 1);
+        return;
+      case 'Escape':
+        close();
+        return;
+      case 'Tab':
+        close();
+        return;
+      default:
+        return;
+    }
+  };
+
+  const focusItem = (index: number) => {
```

Comment:
This is fine as is, however I'm not a huge fan of mutating a parameter in a function.  We should use a local variable to store any updates, just as a general rule.
I changed it to a ternary, but you can modify however you'd like:
```js
const focusItem = (index: number) => {
  const numItems = items.length;
  const idx = index >= numItems ? 0 : index < 0 ? numItems - 1 : index;
  document.getElementById(`${items[idx].id}`).focus();
};
```

Reply:
Updated